configtype/time.go

```diff
@@ -0,0 +1,141 @@
+package configtype
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/invopop/jsonschema"
+)
+
+type timeType int
+
+const (
+	timeTypeZero timeType = iota
+	timeTypeFixed
+	timeTypeRelative
+)
+
+// Time is a wrapper around time.Time that should be used in config
+// when a time type is required. We wrap the time.Time type so that
+// the spec can be extended in the future to support other types of times
+type Time struct {
+	typ      timeType
+	time     time.Time
+	duration time.Duration
+}
+
+func NewTime(t time.Time) Time {
+	return Time{
+		typ:  timeTypeFixed,
+		time: t,
+	}
+}
+
+func NewRelativeTime(d time.Duration) Time {
+	return Time{
+		typ:      timeTypeRelative,
+		duration: d,
+	}
+}
+
+var (
+	timeRFC3339Pattern = `^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(.(\d{1,9}))?(Z|((-|\+)\d{2}:\d{2}))$`
+	timeRFC3339Regexp  = regexp.MustCompile(timeRFC3339Pattern)
+
+	datePattern = `^\d{4}-\d{2}-\d{2}$`
+	dateRegexp  = regexp.MustCompile(datePattern)
+
+	timePattern = patternCases(timeRFC3339Pattern, datePattern, durationPattern)
+)
+
+func (Time) JSONSchema() *jsonschema.Schema {
+	return &jsonschema.Schema{
+		Type:    "string",
+		Pattern: timePattern,
+		Title:   "CloudQuery configtype.Time",
+	}
+}
+
+func (t *Time) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	var err error
+	switch {
+	case timeRFC3339Regexp.MatchString(s):
```

Comment:
Instead of using regular expressions can't we try to parse with each format, then return an error if all formats failed to parse?

Reply:
Yeah, although we still need regexes for the jsonschema definitions